File: crack-data-structures-and-algorithms/leetcode/3sum_q15.cpp
#include <algorithm>
#include <vector>

using namespace std;

// 核心思路
// 题目要求的是数字组合而不是下标，所以可以排序，数组有序是处理最内层两个数的前提条件。
// 1) 首先固定最外层的数字，遍历范围是 [0, |nums|-2)
//    同时要考虑避免重复，所以当最外层数字**至少**是第二轮时，就要检测当前数字是否和前一个相同
//    如果相同就换下一个，直到找到一个不同的
// 2) 三个数的情况只有两层，所以第二层就是最内层，直接用双指针法来检测元素和
//    遍历范围是 [i+1, |nums|)，其中i是外层元素的位置，这里就是第一层
//    最内层同样也要考虑重复的情况：当元素和满足题目要求时，分别对l和r进行相同元素跳过处理
//    注意要先各自移位再进入判断逻辑，因为我们要先**结束**当前case，下一个case才能判断是否和上一个相同
//    判断时别忘了 l < r 的前置要求
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // Precondition.
        if (nums.size() < 3) {
            return {};
        }

        vector<vector<int>> result;
        sort(nums.begin(), nums.end());

        for (auto i = 0; i < nums.size() - 2; ++i) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }

            auto l = i + 1;
            auto r = nums.size() - 1;
            while (l < r) {
                auto sum = nums[i] + nums[l] + nums[r];
                if (sum > 0) {
                    --r;
                } else if (sum < 0) {
                    ++l;
                } else {
                    result.push_back({nums[i], nums[l], nums[r]});

                    do {
                        ++l;
                    } while (l < r && nums[l] == nums[l-1]);

                    do {
                        --r;
                    } while (l < r && nums[r] == nums[r+1]);
                }
            }
        }

        return result;
    }
};

File: crack-data-structures-and-algorithms/leetcode/python-impl/3sum_q15.py
# -*- coding: utf-8 -*-
# 0xCCCCCCCC

# Given an array nums of n integers, are there elements a, b, c in nums such that
# a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

def three_sum(nums):
    """
    :type nums: List[int]
    :rtype: List[List[int]]
    """
    # sort first; taken O(nlogn)
    # iterate from left to right to fix a, then turn this problem to two-sum
    # total taken O(n^2)
    # need to remove duplicate triplets
    nums.sort()
    results = []
    i = 0
    while i < len(nums) - 2:
        a = nums[i]
        if a > 0:
            break
        sum = -a
        bc = two_sum_pairs(nums[i+1:], sum)
        # Found match pairs
        if len(bc) != 0:
            for p in bc:
                results.append([a, p[0], p[1]])
            i += 1
            while i < len(nums) - 2 and nums[i-1] == nums[i]:
                i += 1
        else:
            i += 1

    return results


def two_sum_pairs(nums, val):
    l = 0
    r = len(nums) - 1
    pairs = []
    while l < r:
        sum = nums[l] + nums[r]
        if sum == val:
            pairs.append([nums[l], nums[r]])
            l += 1
            r -= 1
            # Since we just found a match, and if new l or new r equals to the old one
            # skip it becuase it must be duplicates.
            while l < r and nums[l-1] == nums[l]:
                l += 1
            while l < r and nums[r+1] == nums[r]:
                r -= 1
        else:
            if sum < val:
                l += 1
            else:
                r -= 1
    return pairs

nums = [-1, 0, 1, 2, -1, -4]
print(three_sum(nums))
nums = [-1, -1, 0, 1, 1]
print(three_sum(nums))
nums = [0,0,0,0]
print(three_sum(nums))
nums = [-2,0,1,1,2]
print(three_sum(nums))